
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://felixkrueger.github.io/Bismark/bismark/methylation_extraction/">
      
      
        <link rel="prev" href="../deduplication/">
      
      
        <link rel="next" href="../coverage_report/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.49">
    
    
      
        <title>Methylation extraction - Bismark</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.6f8fc17f.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Noto+Sans:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Noto Sans";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
      
        <meta  property="og:type"  content="website" >
      
        <meta  property="og:title"  content="Methylation extraction - Bismark" >
      
        <meta  property="og:description"  content="None" >
      
        <meta  property="og:image"  content="https://felixkrueger.github.io/Bismark/assets/images/social/bismark/methylation_extraction.png" >
      
        <meta  property="og:image:type"  content="image/png" >
      
        <meta  property="og:image:width"  content="1200" >
      
        <meta  property="og:image:height"  content="630" >
      
        <meta  property="og:url"  content="https://felixkrueger.github.io/Bismark/bismark/methylation_extraction/" >
      
        <meta  name="twitter:card"  content="summary_large_image" >
      
        <meta  name="twitter:title"  content="Methylation extraction - Bismark" >
      
        <meta  name="twitter:description"  content="None" >
      
        <meta  name="twitter:image"  content="https://felixkrueger.github.io/Bismark/assets/images/social/bismark/methylation_extraction.png" >
      
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#methylation-extraction" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Bismark" class="md-header__button md-logo" aria-label="Bismark" data-md-component="logo">
      
  <img src="../../images/bismark.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bismark
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Methylation extraction
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/FelixKrueger/Bismark" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    FelixKrueger/Bismark
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../quick_reference/" class="md-tabs__link">
        
  
    
  
  Quick Reference

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../" class="md-tabs__link">
          
  
  Usage

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../options/genome_preparation/" class="md-tabs__link">
          
  
  Full list of options

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../faq/" class="md-tabs__link">
          
  
  FAQ

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Bismark" class="md-nav__button md-logo" aria-label="Bismark" data-md-component="logo">
      
  <img src="../../images/bismark.png" alt="logo">

    </a>
    Bismark
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/FelixKrueger/Bismark" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    FelixKrueger/Bismark
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../quick_reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quick Reference
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Usage
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Usage
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running Bismark
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../genome_preparation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Genome preparation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../alignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Alignment
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../deduplication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Deduplication
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Methylation extraction
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Methylation extraction
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-methylation-extractor-output-looks-like-this-tab-separated" class="md-nav__link">
    <span class="md-ellipsis">
      The methylation extractor output looks like this (tab separated):
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#methylation-extractor-output" class="md-nav__link">
    <span class="md-ellipsis">
      Methylation extractor output
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#strand-specific-methylation-output-files-default" class="md-nav__link">
    <span class="md-ellipsis">
      Strand-specific methylation output files (default):
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#context-dependent-methylation-output-files-comprehensive-option" class="md-nav__link">
    <span class="md-ellipsis">
      Context-dependent methylation output files (--comprehensive option):
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#optional-bedgraph-output" class="md-nav__link">
    <span class="md-ellipsis">
      Optional bedGraph output
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#optional-genome-wide-cytosine-report-output" class="md-nav__link">
    <span class="md-ellipsis">
      (Optional): genome-wide cytosine report output
    </span>
  </a>
  
    <nav class="md-nav" aria-label="(Optional): genome-wide cytosine report output">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gpc-methylation-assessment-context-specific-methylation-summary-report" class="md-nav__link">
    <span class="md-ellipsis">
      GpC methylation assessment: Context-specific methylation summary report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#optional-nome-seq-or-scnmt-seq" class="md-nav__link">
    <span class="md-ellipsis">
      (Optional): NOMe-seq or scNMT-seq
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#m-bias-plot" class="md-nav__link">
    <span class="md-ellipsis">
      M-bias plot
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#iii-running-bismark_methylation_extractor" class="md-nav__link">
    <span class="md-ellipsis">
      (III) Running bismark_methylation_extractor
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../coverage_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Nucleotide coverage report
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../processing_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Processing report
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../summary_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary report
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../filter_nonconverted_reads/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Filtering out non-bisulfite converted reads
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../library_types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Library types
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../concordance/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Concordance across reads
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Full list of options
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Full list of options
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../options/genome_preparation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Genome preparation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../options/alignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Alignment
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../options/methylation_extraction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Methylation extraction
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    FAQ
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            FAQ
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../faq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Bismark FAQs
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../faq/single_cell_pbat/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Single cell PBAT
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../faq/low_mapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Low mapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../faq/context_change/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Context change
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../faq/conversion_efficiency/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Conversion efficiency
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-methylation-extractor-output-looks-like-this-tab-separated" class="md-nav__link">
    <span class="md-ellipsis">
      The methylation extractor output looks like this (tab separated):
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#methylation-extractor-output" class="md-nav__link">
    <span class="md-ellipsis">
      Methylation extractor output
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#strand-specific-methylation-output-files-default" class="md-nav__link">
    <span class="md-ellipsis">
      Strand-specific methylation output files (default):
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#context-dependent-methylation-output-files-comprehensive-option" class="md-nav__link">
    <span class="md-ellipsis">
      Context-dependent methylation output files (--comprehensive option):
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#optional-bedgraph-output" class="md-nav__link">
    <span class="md-ellipsis">
      Optional bedGraph output
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#optional-genome-wide-cytosine-report-output" class="md-nav__link">
    <span class="md-ellipsis">
      (Optional): genome-wide cytosine report output
    </span>
  </a>
  
    <nav class="md-nav" aria-label="(Optional): genome-wide cytosine report output">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gpc-methylation-assessment-context-specific-methylation-summary-report" class="md-nav__link">
    <span class="md-ellipsis">
      GpC methylation assessment: Context-specific methylation summary report
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#optional-nome-seq-or-scnmt-seq" class="md-nav__link">
    <span class="md-ellipsis">
      (Optional): NOMe-seq or scNMT-seq
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#m-bias-plot" class="md-nav__link">
    <span class="md-ellipsis">
      M-bias plot
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#iii-running-bismark_methylation_extractor" class="md-nav__link">
    <span class="md-ellipsis">
      (III) Running bismark_methylation_extractor
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="methylation-extraction">Methylation extraction</h1>
<p>Bismark comes with a supplementary <code>bismark_methylation_extractor</code> script which operates on Bismark result files and extracts the methylation call for every single C analysed. The position of every single C will be written out to a new output file, depending on its context (CpG, CHG or CHH), whereby methylated Cs will be labelled as forward reads (+), non-methylated Cs as reverse reads (-). The resulting files can be imported into a genome viewer such as SeqMonk (using the generic text import filter) and the analysis of methylation data can commence. Alternatively, the output of the methylation extractor can be transformed into a <code>bedGraph</code> and <code>coverage</code> file using the option <code>--bedGraph</code> (see also <code>--counts</code>). This step can also be accomplished from the methylation extractor output using the stand-alone script <code>bismark2bedGraph</code> (also part of the Bismark package available for download at <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/">bioinformatics.babraham.ac.uk</a>). The <code>coverage</code> file can also be imported into SeqMonk directly using <code>Import Data &gt; Bismark (cov)</code>. Optionally, the bedGraph counts output can be used to generate a genome-wide cytosine report which reports the number on every single CpG (optionally every single cytosine) in the genome, irrespective of whether it was covered by any reads or not. As this type of report is informative for cytosines on both strands the output may be fairly large (~46mn CpG positions or &gt;1.2bn total cytosine positions in the human genome...). The bedGraph to genome-wide cytosine report conversion can also be run individually using the stand- alone module <code>coverage2cytosine</code> (also part of the Bismark package available for download at <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/">bioinformatics.babraham.ac.uk</a>).</p>
<p>As of Bismark version 0.6 or higher the default input format for the <code>bismark_methylation_extractor</code> is BAM/SAM (or potentially CRAM if you’ve got Samtools 1.2+ installed).</p>
<h3 id="the-methylation-extractor-output-looks-like-this-tab-separated">The methylation extractor output looks like this (tab separated):</h3>
<ol>
<li>seq-ID</li>
<li>methylation state</li>
<li>chromosome</li>
<li>start position <em>(= end position)</em></li>
<li>methylation call</li>
</ol>
<p>Methylated cytosines receive a <code>+</code> orientation, unmethylated cytosines receive a <code>-</code> orientation.</p>
<p>Examples for cytosines in CpG context:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>HWUSI-EAS611_0006:3:1:1058:15806#0/1 - 6 91793279 z
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>HWUSI-EAS611_0006:3:1:1058:17564#0/1 + 8 122855484 Z
</code></pre></div>
<p>Examples for cytosines in CHG context:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>HWUSI-EAS611_0006:3:1:1054:1405#0/1 - 7 89920171 x
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>HWUSI-EAS611_0006:3:1:1054:1405#0/1 + 7 89920172 X
</code></pre></div>
<p>Examples for cytosines in CHH context:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>HWUSI-EAS611_0006:3:1:1054:1405#0/1 - 7 89920184 h
</code></pre></div>
<p>The <code>bismark_methylation_extractor</code> comes with a few options, such as ignoring the first <int> number of positions in the methylation call string, e.g. to remove a restriction enzyme site (if RRBS is performed with non-directional BS-Seq libraries it might be required to remove reconstituted MspI sites at the beginning of each read as they will introduce a bias into the first methylation call). Another useful option for paired-end reads is called <code>--no_overlap</code> (on by default): specifying this option will extract the methylation calls of overlapping parts in the middle of paired-end reads only once (using the calls from the first read which is presumably the one with a lowest error rate).</p>
<p>For a full list of options type: <code>bismark_methylation_extractor --help</code> on the command line or refer to the Appendix section at the end of this User Guide.</p>
<h3 id="methylation-extractor-output">Methylation extractor output</h3>
<p>By default, the <code>bismark_methylation_extractor</code> discriminates between cytosines in CpG, CHG or CHH context. If desired, CHG and CHH contexts can be merged into a single non-CpG context by specifying the option <code>--merge_non_CpG</code> (as a word of warning, this might produce files with up to several hundred million lines...).</p>
<h3 id="strand-specific-methylation-output-files-default">Strand-specific methylation output files (default):</h3>
<p>As its default option, the <code>bismark_methylation_extractor</code> will produce a strand-specific output which will use the following abbreviations in the output file name to indicate the strand the alignment came from:</p>
<div class="highlight"><pre><span></span><code>OT    –  original top strand
CTOT  –  complementary to original top strand
OB    –  original bottom strand
CTOB  –  complementary to original bottom strand
</code></pre></div>
<p>Methylation calls from OT and CTOT will be informative for cytosine methylation positions on the original top strand, calls from OB and CTOB will be informative for cytosine methylation positions on the original bottom strand. Please note that specifying the <code>--directional</code> (the default mode) option in the Bismark alignment step will not report any alignments to the CTOT or CTOB strands.</p>
<p>As cytosines can exist in any of three different sequence contexts (CpG, CHG or CHH) the <code>bismark_methylation_extractor</code> default output will consist of 12 individual output files per input file (<code>CpG_OT_...</code>, <code>CpG_CTOT_...</code>, <code>CpG_OB_...</code> etc.).</p>
<h3 id="context-dependent-methylation-output-files-comprehensive-option">Context-dependent methylation output files (<code>--comprehensive</code> option):</h3>
<p>If strand-specific methylation is not of interest, all available methylation information can be pooled into a single context-dependent file (information from any of the four strands will be pooled). This will default to three output files (CpG-context, CHG-context and CHH-context), or result in 2 output files (CpG-context and Non-CpG-context) if <code>--merge_non_CpG</code> was selected (note that this can result in enormous file sizes for the non-CpG output).</p>
<p>Both strand-specific and context-dependent options can be combined with the <code>--merge_non_CpG</code> option.</p>
<h3 id="optional-bedgraph-output">Optional bedGraph output</h3>
<p>The Bismark methylation extractor can optionally also output a file in <a href="http://genome.ucsc.edu/goldenPath/help/bedgraph.html"><code>bedGraph</code></a> format which uses 0-based genomic start and 1- based end coordinates. The module <code>bismark2bedGraph</code> (part of the Bismark package) may also be run individually. It will be sorted by chromosomal coordinates and looks like this:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>&lt;chromosome&gt; &lt;start position&gt; &lt;end position&gt; &lt;methylation percentage&gt;
</code></pre></div>
<p>As the methylation percentage is <em>per se</em> not informative of the actual read coverage of detected methylated or unmethylated reads at a position, <code>bismark2bedGraph</code> also writes out a coverage file (using 1-based genomic genomic coordinates) that features two additional columns:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>&lt;chromosome&gt; &lt;start position&gt; &lt;end position&gt; &lt;methylation percentage&gt; &lt;count methylated&gt; &lt;count unmethylated&gt;
</code></pre></div>
<p>These two additional columns enable basically any downstream processing from the file. By default, this mode will only consider cytosines in CpG context, but it can be extended to cytosines in any sequence context by using the option <code>--CX</code> (cf. Appendix (III)).</p>
<h3 id="optional-genome-wide-cytosine-report-output">(Optional): genome-wide cytosine report output</h3>
<p>Starting from the <code>coverage</code> output, the Bismark methylation extractor can optionally also output a genome-wide cytosine methylation report. The module <code>coverage2cytosine</code> (part of the Bismark package) may also be run individually. It is also sorted by chromosomal coordinates but also contains the sequence context and is in the following format:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>&lt;chromosome&gt; &lt;position&gt; &lt;strand&gt; &lt;count methylated&gt; &lt;count unmethylated&gt; &lt;C-context&gt; &lt;trinucleotide context&gt;
</code></pre></div>
<p>The main difference to the <code>bedGraph</code> or <code>coverage</code> output is that <strong>every</strong> cytosine on both the top and bottom strands will be considered irrespective of whether they were actually covered by any reads in the experiment or not. For this to work one has to also specify the genome that was used for the <code>Bismark</code> alignments using the option <code>--genome_folder &lt;path&gt;</code>. As for the <code>bedGraph</code> mode, this will only consider cytosines in CpG context by default but can be extended to cytosines in any sequence context by using the option <code>--CX</code> (cf. Appendix (III)). Be aware though that this might mean an output with individual lines for more than 1.1 billion cytosines for any large mammalian genome...</p>
<h4 id="gpc-methylation-assessment-context-specific-methylation-summary-report">GpC methylation assessment: Context-specific methylation summary report</h4>
<p>Every time <code>coverage2cytosine</code> is run, the counts for each cytosine context are recorded and printed to a special file called <code>*.cytosine_context_summary.txt</code> for each input coverage file. </p>
<p>The report (<a href="https://github.com/FelixKrueger/Bismark/issues/321">introduced here</a>) looks at 2 bp downstream, as well as 1 bp upstream of the cytosine taking part in the methylation call. This is useful for looking at methylation in specific contexts (e.g. <code>CpA</code> only), and also when using <code>GpC</code> methylases that introduce methylation in <code>GpC</code> context. The report looks like this: </p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>upstream    C-context   full context    count methylated    count unmethylated  percent methylation
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>A   CAA ACAA    308398  93360   23.24
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>C   CAA CCAA    329555  75518   18.64
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>G   CAA GCAA    242262  118175  32.79
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a>T   CAA TCAA    302209  86219   22.20
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a>A   CAC ACAC    266786  91121   25.46
<a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a>C   CAC CCAC    409749  89154   17.87
<a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a>...
</code></pre></div>
<h4 id="optional-nome-seq-or-scnmt-seq">(Optional): NOMe-seq or scNMT-seq</h4>
<p>The <code>coverage2cytosine</code> module can be instructed that a sample is a NOMe-seq (<a href="https://genome.cshlp.org/content/22/12/2497.long"><strong>N</strong>ucleosome <strong>O</strong>ccupancy and <strong>Me</strong>thylome sequencing</a>) or scNMT-seq (<a href="https://www.nature.com/articles/s41467-018-03149-4"><strong>s</strong>ingle-<strong>c</strong>ell <strong>N</strong>ucleosome, <strong>M</strong>ethylation and <strong>T</strong>ranscription sequencing</a>) sample, where accessible DNA gets methylated in a GpC context (sets option <code>--gc</code> as well). The option <code>--nome-seq</code>:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a> (i) filters the genome-wide CpG-report to only output cytosines in ACG and TCG context
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>(ii) filters the GC context output to only report cytosines in GCA, GCC and GCT context
</code></pre></div>
<p>Both of these measures aim to reduce unwanted biases, i.e. the influence of <code>G-CG</code> (intended) and <code>C-CG</code> (off-target) on endogenous CpG methylation, and the influence of CpG methylation on (the NOMe-seq specific) <code>GC</code> context methylation. <strong>PLEASE NOTE</strong> that NOMe-seq data requires a <code>.cov.gz</code> file as input which has been generated in non-CG mode!! (<code>--CX</code>), else the <code>GpC</code> output file will be empty...</p>
<h3 id="m-bias-plot">M-bias plot</h3>
<p>Starting with Bismark v0.8.0, the Bismark methylation extractor also produces a methylation bias plot which shows the methylation proportion across each possible position in the read (described in further detail in: <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-10-r83">Hansen et al., Genome Biology, 2012, 13:R83</a>). The data for the M-bias plot is also written into a coverage text file (ending in <code>.cov</code> or <code>.cov.gz</code>) and is in the following format:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>&lt;read position&gt; &lt;count methylated&gt; &lt;count unmethylated&gt; &lt;% methylation&gt; &lt;total coverage&gt;
</code></pre></div>
<p>This allows generating nice graphs by alternative means, e.g. using R or Excel. The plot is also drawn into a .png file which requires the Perl module GD::Graph (more specifically, both modules GD::Graph::lines and GD::Graph::colour are required); if GD::Graph cannot be found on the system, only the table will be printed. The plot also contains the absolute number of methylation calls (both methylated and unmethylated) per position. For paired-end reads two individual M-bias plots will be drawn.</p>
<p>The M-bias plot can for example show the methylation bias at the start of reads in PBAT-Seq experiments:</p>
<p><img alt="SE PBAT" src="../../images/PBAT_SE_M-bias.jpg" /></p>
<p>For more on this topic please also see <a href="https://sequencing.qcfail.com/articles/mispriming-in-pbat-libraries-causes-methylation-bias-and-poor-mapping-efficiencies/">this post on QCFail.com</a>.</p>
<p>Or it can reveal a 3’-end-repair bias at the first couple of positions in read 2 of paired-end reads, like here:</p>
<p><img alt="PE fill-in" src="../../images/PE_M-bias.jpg" /></p>
<p>For more on this topic please also see <a href="https://sequencing.qcfail.com/articles/library-end-repair-reaction-introduces-methylation-biases-in-paired-end-pe-bisulfite-seq-applications/">this post on QCFail.com</a>.</p>
<p>The M-bias plot should enable researchers to make an informed decision whether or not to leave the bias in the final data or to remove it (e.g. using the methylation extractor option <code>--ignore</code>).</p>
<h3 id="iii-running-bismark_methylation_extractor">(III) Running <code>bismark_methylation_extractor</code></h3>
<p><strong>USAGE:</strong> <code>bismark_methylation_extractor [options] &lt;filenames&gt;</code></p>
<p>A typical command for a single-end file could look like this:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>bismark_methylation_extractor --gzip sample_bismark_bt2.bam
</code></pre></div>
<p>A typical command for a paired-end file could look like this:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>bismark_methylation_extractor --gzip sample_bismark_bt2_pe.bam
</code></pre></div>
<p>The methylation extractor can also auto-detect the alignment mode and will set the options above automatically. A typical command including the optional <code>bedGraph</code> output could look like this:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>bismark_methylation_extractor --gzip --bedGraph --buffer_size 10G sample_bismark_bt2.bam
</code></pre></div>
<p>A typical command including the optional genome-wide cytosine methylation report could look like this:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>bismark_methylation_extractor --gzip --bedGraph --buffer_size 10G --cytosine_report --genome_folder /path_to_genome_folder/ sample_bismark_bt2.bam
</code></pre></div>







  
  






                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tabs"], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.88dd0f4e.min.js"></script>
      
    
  </body>
</html>